File: part1/section2/semantics_demo.py
# python: type system, closures, and numeric model

def closure_capture():
    # late binding: all closures share the same variable cell
    fns_late = []
    for i in range(3):
        fns_late.append(lambda: i)
    # default-argument idiom forces capture by value at definition time
    fns_early = []
    for i in range(3):
        fns_early.append(lambda i=i: i)
    return [f() for f in fns_late], [f() for f in fns_early]

def type_system():
    results = []
    x = 5
    results.append(("rebind int to str", type(x).__name__))
    x = "five"
    results.append(("after rebinding", type(x).__name__))
    try:
        _ = "5" + 3
    except TypeError as e:
        results.append(("'5' + 3", "TypeError: " + str(e)))
    results.append(("5 + 3.0", repr(5 + 3.0)))
    return results

def numeric_model():
    # arbitrary precision: no overflow
    big = 2 ** 63
    return big, big * big, (2**31 - 1) + 1

if __name__ == "__main__":
    late, early = closure_capture()
    print("closures (late binding):  ", late)
    print("closures (default arg):   ", early)
    print()
    for label, val in type_system():
        print(f"  {label:22} -> {val}")
    print()
    b, bsq, edge = numeric_model()
    print("2**63          =", b)
    print("(2**63)**2     =", bsq)
    print("(2**31-1)+1    =", edge)

File: part1/section2/semantics_demo.js
// javascript: type system, closures, and numeric model

function closureCapture() {
    // var is function-scoped: all three closures share one binding
    const fnsVar = [];
    for (var i = 0; i < 3; i++) {
        fnsVar.push(() => i);
    }
    // let is block-scoped: a fresh binding is created per iteration
    const fnsLet = [];
    for (let j = 0; j < 3; j++) {
        fnsLet.push(() => j);
    }
    return [fnsVar.map(f => f()), fnsLet.map(f => f())];
}

function typeSystem() {
    return [
        ["'5' + 3",        JSON.stringify("5" + 3)],
        ["'5' - 3",        JSON.stringify("5" - 3)],
        ["[] + {}",        JSON.stringify([] + {})],
        ["1 == '1'",       String(1 == "1")],
        ["1 === '1'",      String(1 === "1")],
        ["typeof (5)",     typeof 5],
        ["typeof ('five')", typeof "five"],
    ];
}

function numericModel() {
    return {
        maxSafe: Number.MAX_SAFE_INTEGER,
        overflow: Number.MAX_SAFE_INTEGER + 2,
        int32wrap: (2147483647 | 0) + 1,
        pointOne: 0.1 + 0.2,
    };
}

const [varRes, letRes] = closureCapture();
console.log("closures (var):", varRes);
console.log("closures (let):", letRes);
console.log();
for (const [label, val] of typeSystem()) {
    console.log("  " + label.padEnd(22) + " -> " + val);
}
console.log();
const n = numericModel();
console.log("MAX_SAFE_INTEGER      =", n.maxSafe);
console.log("MAX_SAFE_INTEGER + 2  =", n.overflow);
console.log("(2147483647|0) + 1    =", n.int32wrap);
console.log("0.1 + 0.2             =", n.pointOne);
